src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -79,11 +87,48 @@ enum Signal {
    */
   final Map<RepoRecordedInput, String> recordedInputValues = new TreeMap<>();
 
+  RepoFetchingSkyKeyComputeState(String repoName) {
+    this.repoName = repoName;
+    reset();
+  }
+
+  // This may only be called from the host Skyframe thread, *and* only when no worker thread is
+  // running.
+  private void reset() {
+    workerExecutorService =
+        MoreExecutors.listeningDecorator(
+            Executors.newThreadPerTaskExecutor(
+                Thread.ofVirtual().name("starlark-repository-" + repoName).factory()));
+    signalSemaphore.drainPermits();
+    delegateEnvQueue.clear();
+    recordedInputValues.clear();
+  }
+
+  /**
+   * Releases a permit on the {@code signalSemaphore} and immediately expect a fresh Environment
+   * back. This may only be called from the worker thread.
+   */
   SkyFunction.Environment signalForFreshEnv() throws InterruptedException {
-    signalQueue.put(Signal.RESTART);
+    signalSemaphore.release();
     return delegateEnvQueue.take();
   }
 
+  /**
+   * Starts a worker thread running the given callable. This sets the {@code workerFuture} field,
+   * and makes sure to release a permit on the {@code signalSemaphore} when the worker finishes,
+   * successfully or otherwise. Returns the worker future. This may only be called from the host
+   * Skyframe thread.
+   */
+  ListenableFuture<RepositoryDirectoryValue.Builder> startWorker(
+      Callable<RepositoryDirectoryValue.Builder> c) {
+    var workerFuture = workerExecutorService.submit(c);
+    this.workerFuture = workerFuture;
+    workerFuture.addListener(signalSemaphore::release, directExecutor());
+    return workerFuture;
+  }
+
+  // This may be called from any thread, including the host Skyframe thread and the
+  // high-memory-pressure listener thread.
   @Override
   public void close() {
     var myWorkerFuture = workerFuture;
```

Comment:
if the worker thread gets 'orphaned' while the host Skyframe thread is active, then the host would call `state.closeAndWaitForTermination()`, ensuring the worker thread is joined. But if it gets 'orphaned' when the host Skyframe thread is _inactive_ (ie. having returned `null` but not yet restarted), then you're right the worker thread would live on for an indefinite amount of time.

I added a clause to guard against that specific case (so when the Skyframe thread becomes active again, the first thing it does is to join the worker thread).

Also added a TODO to reconsider the ban on blocking work in `close()`...

Reply:
My $.02: can we not have another complicated dance? The documentation of `close()` doesn't say *why* it shouldn't perform any blocking operations and it looks silly that a stray thread is preferred to blocking a bit.

I'd much rather not have to think about what happens in odd cases likewhen the Skyframe evaluation is stopped due to an independent error (or a SIGINT) before the worker thread quits and another Bazel command is issued right after that, etc.

Reply:
> The documentation of close() doesn't say why it shouldn't perform any blocking operations

Presumably because it can be called from any thread. It's a bit scary to perform arbitrary blocking operations in random threads. If we want to reverse the decision, I'd rather get the Core team's input, and I suggest that we do this in a follow-up instead of right away in this PR, which is getting a bit big.

Reply:
Well, but at the same time, it's pretty scary to have orphaned threads doing I/O. But then again, that's arguably not for this pull request; you do cancel the future, which is supposed to kill off the now-orphan thread very quickly.